gio-macros/tests/actions.rs

```diff
@@ -0,0 +1,79 @@
+use gio::prelude::*;
+use std::cell::RefCell;
+use std::rc::Rc;
+
+#[derive(glib::Downgrade)]
+pub struct MyActionGroup {
+    group: gio::SimpleActionGroup,
+    log: Rc<RefCell<Vec<String>>>,
+}
+
+impl MyActionGroup {
+    pub fn create() -> Self {
+        let app = Self {
+            group: gio::SimpleActionGroup::new(),
+            log: Default::default(),
+        };
+        app.register_actions(&app.group);
+        app
+    }
+}
+
+#[gio::actions]
+impl MyActionGroup {
```

Comment:
Is this really needed? I would rather have everything on the same Impl and decorate all the actions with `#[action...]`

Reply:
I am afraid it is not possible. In this case proc-macro will not be able to generate a single method which gathers all actions into an ActionMap.

Reply:
You don't need to gather the actions if you assume that self is a `glib::IsA<gio::ActionMap>` then, you can directly add the actions without requiring an extra method. Of course that would only work for widgets/object that implement ActionMap. Which means gtk::Application & gtk::ApplicationWindow for now.  We can then have a proc macro that would automatically implement ActionMap on top of a gobject subclass which would look much cleaner API wise I guess.

Reply:
> Of course that would only work for widgets/object that implement ActionMap.

Or something that implements `Deref<Target=ActionMap>` :)

Reply:
1. Requiring that `Self` implements `ActionMap` makes this macro useless for my personal projects where I do not use inheritance but newtypes to create abstractions.
2. Even if `Self: ActionMap` is assumed, it doesn't help. Code which generates an action should be invoked from somewhere. Attribute proc-macro attached to `impl` item allows to append a single method (`register_actions` in this case) and then this method can be called somewhere. Attribute proc-macro attached to each action handler individually will not be able to do so. It can generate another function and that's it. There is no way to inject *invocation* of these functions (Unless we assume additional requirement. `GtkWidget::created-event` would fit here if existed :laughing: )

Reply:
>     1. Requiring that `Self` implements `ActionMap` makes this macro useless for my personal projects where I do not use inheritance but newtypes to create abstractions.

Sure thing, the reason why everyone went that way when gtk-rs apps started showing up is that subclassing wasn't possible (required a ton of boilerplate). It still requires a bit of boilerplate and such macro would help in that direction :)  

>     2. Even if `Self: ActionMap` is assumed, it doesn't help. Code which generates an action should be invoked from somewhere. Attribute proc-macro attached to `impl` item allows to append a single method (`register_actions` in this case) and then this method can be called somewhere. Attribute proc-macro attached to each action handler individually will not be able to do so. It can generate another function and that's it. There is no way to inject _invocation_ of these functions (Unless we assume additional requirement. `GtkWidget::created-event` would fit here if existed laughing )

I don't see why you would need to register them, if the #[action] is called on a function already on self, you have access to self, so you can just do self.add_action directly instead of storing them somewhere, no?

Reply:
@bilelmoussaoui Who should call `self.add_action`? Or, where should macro put that `self.add_action`?

Reply:
Ah indeed. We can have a trait (see the composite template macro) with a register_actions method that the user would have to call when the object is `constructed`. Until we have some kind of way to automatically hook those as the user could be using a combo of different macros (composite template, the actions one, the gobject properties one, something else in the future?)

Reply:
Proc-macros are **stateless** and they cannot generate **one** `register_action` method (or one implementation of some trait) while being invoked **multiple** times (once per annotation).

Reply:
Here's a rough pseudo code idea of what I had in mind
```rust
mod imp {

    #[derive(CompositeTemplate, Actions)]
    #[actions(group="widget")]
    pub struct Widget {

    }

    impl ObjectSubclass for Widget {
        ....

        fn type_init(type_: subclass::InitializedType<Self>) {
            type_.add_interface::<gio::ActionMap>();
            type_.add_interface::<gio::ActionGrouo>();
        }
    }

    impl ObjectImpl for Widget {
        fn constructed(&self, obj: &Self::Type) {
            obj.register_actions();
        }
    }

    # both those traits would be automatically implemented when using #[actions(group....)]
    impl ActionMap for Widget {}
    impl ActionGroup for Widget {}
}

glib::wrapper! {
    pub struct Widget(ObjectSubclass<imp::Widget>) @derive gtk::Widget, @implements gio::ActionMap, gio::ActionGroup;
}


impl widget {

    #[action]
    fn name(&self, action: &gio::Action, target: Option<&glib::Value>) {

    }
}
```

Reply:
```rust
    #[action]
    fn name(&self, action: &gio::Action, target: Option<&glib::Value>) {

    }
```

What a macro should emit here? Notice, that it is **inside** of an `impl` block and such block can contain functions or constants only. Nothing else.

Reply:
Does it have to register anything actually? In my idea it was mostly used to identify the action functions so that you can register them later

Reply:
Of course, it should emit something. At least the function itself :laughing: with **stripped out** attribute (otherwise it won't compile).

Reply:
Ah yeah, that's what I meant, as in not do anything else other than what the function contains ^^